### PR TITLE
Fixes to initialization order and circular dependencies

### PR DESCRIPTION
**Pull request by rgknox**
### Description:

API change, synced with: https://github.com/ESCOMP/CTSM/pull/1007

FatesInterfaceMod held both the routines, and also the data required to communicate both with the HLM code, and with FATES code.  This created circular dependencies, where fates modules needed to know about the FatesInterfaceMod data structures (because they use them), but couldn't becuause the FatesInterfaceMod routines wanted to know about the various fates routines. 

The solution here is to split the data structures out of FatesInterfaceMod, and give them their own file.  This file has very low dependency on other files, and how FatesInterfaceMod routines are not bound to it.

This addresses issue #583 and #639 

**This fix is critical in order to progress E3SM to API 10+**. https://github.com/E3SM-Project/E3SM/pull/3563

This should not change answers, but it will avoid diagnostic errors with multiple elements.  This should enable b4b restarts with cohort age functions.

### Collaborators:

@xuchongang 

### Checklist:

- [x] I have updated the in-code documentation .AND. (the [technical note](https://github.com/NGEET/fates-docs) .OR. the wiki) accordingly.
- [x] I have read the [**CONTRIBUTING**](https://github.com/NGEET/fates/blob/master/CONTRIBUTING.md) document.
- [ ] FATES PASS/FAIL regression tests were run

### Test Results:

TBD

CTSM (or) E3SM (specify which) test hash-tag:

CTSM (or) E3SM (specify which) baseline hash-tag:

FATES baseline hash-tag:

Test Output:

### CODE REVIEW

File: main/EDTypesMod.F90 [hunk 1]
```diff
@@ -189,7 +189,7 @@ module EDTypesMod
                                                         ! in PRTGenericMod.F90. examples are carbon12_element
                                                         ! nitrogen_element, etc.
 
-  integer, public :: element_pos(num_organ_types)       ! This is the reverse lookup
+  integer, public :: element_pos(num_element_types)       ! This is the reverse lookup
```
Comment (glemieux) [inline main/EDTypesMod.F90:192]: I see `num_organ_types` and `num_element_types` are defined to have the same value currently.  I'm guessing `num_organ_types` was the misapplied originally?  Or was it something else I'm missing?

File: parteh/PRTLossFluxesMod.F90 [hunk 2]
```diff
@@ -18,8 +18,8 @@ module PRTLossFluxesMod
   use PRTGenericMod, only : check_initialized
   use PRTGenericMod, only : num_organ_types
   use PRTGenericMod, only : prt_global
-  use FatesInterfaceMod, only : hlm_freq_day
-
+  use FatesConstantsMOd, only : years_per_day
```
Comment (glemieux) [inline parteh/PRTLossFluxesMod.F90:21]: Super tiny formatting nitpick here: `FatesConstantsMod`

File: main/FatesInterfaceMod.F90 [hunk 3]
```diff
@@ -9,666 +9,225 @@ module FatesInterfaceMod
...
 
    ! CIME Globals
-   use shr_log_mod         , only : errMsg => shr_log_errMsg
-   use shr_infnan_mod      , only : nan => shr_infnan_nan, assignment(=)
+   use shr_log_mod               , only : errMsg => shr_log_errMsg
+   use shr_infnan_mod            , only : nan => shr_infnan_nan, assignment(=)
 
-   implicit none
-
-   private        ! By default everything is private
```
[the remark this answers is not in the document]
Comment (ekluzek) [reply main/FatesInterfaceMod.F90:49]: Just chiming in to agree with @glemieux. First, implicit none is pretty much required in FORTRAN. It really needs to be there at least at the top of the module. And as Greg says default private is a a really good idea, and you should just declare what is intended to be public. If you don't make it default private everything will be shared with something that uses it. That can be really problematic.

[the remark this answers is not in the document]
Comment (rgknox) [reply main/FatesInterfaceMod.F90:49]: thanks guys, that was an oversite, will correct

File: main/EDTypesMod.F90 (hunk 1 above)
Comment (rgknox) [reply main/EDTypesMod.F90:192]: yes, they just happen to be the same size
It was originally a typo, the number of element positions should have the same size as the number of elements.

File: parteh/PRTLossFluxesMod.F90 (hunk 2 above)
Comment (rgknox) [reply parteh/PRTLossFluxesMod.F90:21]: haha, good catch

File: main/FatesInterfaceMod.F90 [hunk 4]
```diff
@@ -9,666 +9,225 @@ module FatesInterfaceMod
...
-
-      ! These are the boundary conditions that the FATES model returns to its HLM or 
-      ! driver. It has the same allocation strategy and similar vector types.
-      
-      type(bc_out_type), allocatable  :: bc_out(:)
-
-   contains
-      
-      procedure, public :: zero_bcs
-      procedure, public :: set_bcs
```
[the remark this answers is not in the document]
Comment (rgknox) [reply main/FatesInterfaceMod.F90:619]: I found where it was first added, but there is no call to it.  It is not clear exactly what this routine is intneded to do.  Perhaps its helper code that is ready to be deployed once the HLMs have salinity info available (which they don't yet?). We should create an issue. 
FYI, as far as I can tell, we never had a call to this routine.

[the remark this answers is not in the document]
Comment (glemieux) [reply main/FatesInterfaceMod.F90:619]: Roger that, I'll write one up.

[the remark this answers is not in the document]
Comment (glemieux) [reply main/FatesInterfaceMod.F90:619]: Here's the issue: https://github.com/NGEET/fates/issues/648

**Comment by rgknox** [comment]
Testing on cheyenne looks good.
(Note crprnc needed a slight modification and re-build to get these tests working: see here: https://github.com/ESMCI/cime/issues/3525)
*However: I don't suspect we will need the cprnc changes going forward.

/glade/scratch/rgknox/clmed-tests/fates-sci.1.35.2_api.11.0.0-clm5.0.30-Cc7961174-F3bba389b.fates.cheyenne.intel
/glade/scratch/rgknox/clmed-tests/fates-sci.1.35.2_api.11.0.0-clm5.0.30-Cc7961174-F3bba389b.fates.cheyenne.gnu

Here are the differences in the regression tests, and they are expected.  All other variables are the same.


```
cat /glade/scratch/rgknox/ERS_Ld60.f45_f45_mg37.I2000Clm50FatesCruGs.cheyenne_intel.clm-Fates.GC.20200511_123018_8i08wn/run/ERS_Ld60.f45_f45_mg37.I2000Clm50FatesCruGs.cheyenne_intel.clm-Fates.GC.20200511_123018_8i08wn.clm2.h0.0001-01-01-00000.nc.cprnc.out | grep RMS
 RMS fates_levelem                            1            NORMALIZED          0
 RMS fates_elmap_levelpft                     1            NORMALIZED          0
 RMS fates_pftmap_levelpft                    7            NORMALIZED          2
 RMS fates_elmap_levelcwd                     1            NORMALIZED          0
 RMS fates_cwdmap_levelcwd                    2            NORMALIZED          2
 RMS fates_elmap_levelage                     1            NORMALIZED          0
 RMS fates_agemap_levelage                    4            NORMALIZED          2
```

File: parteh/PRTLossFluxesMod.F90 (hunk 2 above)
Comment (rgknox) [reply parteh/PRTLossFluxesMod.F90]: resolved

File: main/FatesInterfaceMod.F90 (hunk 3 above)
[the remark this answers is not in the document]
Comment (rgknox) [reply main/FatesInterfaceMod.F90:49]: resolved

File: main/EDInitMod.F90 [hunk 5]
```diff
@@ -66,7 +66,7 @@ module EDInitMod
   use PRTGenericMod,          only : nitrogen_element
   use PRTGenericMod,          only : phosphorus_element
   use PRTGenericMod,          only : SetState
-  use FatesPlantHydraulicsMod,  only : InitHydroGlobals
+!!  use FatesPlantHydraulicsMod,  only : InitHydroGlobals
   use PRTAllometricCarbonMod, only : InitPRTGlobalAllometricCarbon
```
[the remark this answers is not in the document]
Comment (rgknox) [reply main/EDInitMod.F90]: resolved

**Comment by rgknox** [comment]
re-tested with our gnu tests to verify syntax changes per review requests
This PR is complete on my end @glemieux

File: main/FatesInterfaceMod.F90 (hunk 4 above)
[the remark this answers is not in the document]
Comment (xuchongang) [reply main/FatesInterfaceMod.F90:619]: @rgknox and @glemieux , yes, commit 52d8e34 will use the salinity from the FATES parameter file (it is now begin to be implemented from ATS through ATES-FATES). Not yet for ELM or CLM.

**Pull request merged**